### PR TITLE
Added API to create a custom shell layout

### LINKED ISSUE #1546: Way to specify default view?
The very first time a browser accesses Theia, views like _Files_ or _Git_ are available on the left side, but none are actually open.

image

Is there a way to specify that by default, the _Files_ view should be open in this situation?

### PR DESCRIPTION
**Pull request by spoenemann**
Fixes #1546.

### CODE REVIEW

File: packages/core/src/browser/shell/shell-layout-restorer.ts [hunk 1]
```diff
@@ -101,6 +78,15 @@ export class ShellLayoutRestorer implements CommandContribution {
         }
     }
 
+    async restoreLayout(app: FrontendApplication): Promise<void> {
+        const serializedLayoutData = await this.storageService.getData<string>(this.storageKey);
+        if (serializedLayoutData === undefined) {
+            throw new Error('Layout data is not available.');
```
Comment (akosyakov) [inline packages/core/src/browser/shell/shell-layout-restorer.ts]: Could we use a special error code to filter out real errors from this or have another contract?

File: packages/core/src/browser/frontend-application.ts [hunk 2]
```diff
@@ -170,6 +170,34 @@ export class FrontendApplication {
...
     }
 
+    /**
+     * Initialize the shell layout either using the layout restorer service or, if no layout has
+     * been stored, by creating the default layout.
+     */
+    protected async initializeLayout(): Promise<void> {
+        try {
+            // Try to restore the shell layout from the storage service
+            await this.layoutRestorer.restoreLayout(this);
```
Comment (elaihau) [inline packages/core/src/browser/frontend-application.ts]: Before this change, we have `await app.shell.pendingUpdates;` in the `try-catch` to make sure errors that happen during the restore are caught. Could you please explain what is the equivalence in your change?

File: packages/core/src/browser/frontend-application.ts [hunk 3]
```diff
@@ -170,6 +170,34 @@ export class FrontendApplication {
...
+        } catch (error) {
+            this.logger.info(error.toString());
```
[the remark this answers is not in the document]
Comment (spoenemann) [reply packages/core/src/browser/frontend-application.ts]: > And calling toString is redundant.

No, it's not. Without `toString()` I get
```
Uncaught (in promise) TypeError: loggable is not a function
    at logger.ts:323
```

[the remark this answers is not in the document]
Comment (AlexTugarev) [reply packages/core/src/browser/frontend-application.ts]: `error` can be also `undefined`

File: packages/core/src/browser/shell/shell-layout-restorer.ts (hunk 1 above)
Comment (spoenemann) [reply packages/core/src/browser/shell/shell-layout-restorer.ts]: I'll change `layoutRestorer.restoreLayout()` to return `Promise<boolean>`.

**Comment by spoenemann** [comment]
> I think the extension is capable to initialize the layout without rebinding the frontend application, which is quite invasive, via `initializeLayout` lifecycle hook.

See https://github.com/theia-ide/theia/issues/1546#issuecomment-380022650

**Pull request merged**